cyclonedds/sub.py

```diff
@@ -371,7 +371,7 @@ def wait_for_historical_data(self, timeout: int) -> bool:
         raise DDSException(ret, f"Occured while waiting for historical data in {repr(self)}")
 
     def lookup_instance(self, sample: _T) -> Optional[int]:
-        ret = ddspy_lookup_instance(self._ref, sample.serialize())
+        ret = ddspy_lookup_instance(self._ref, sample.serialize_key())
```

Comment:
I'd expect that this serialize also needs padding `\0`s, similar to `pub.lookup_instance`?

Reply:
Actually, no, and `pub.lookup_instance` doesn't really need it either. This padding business is a bit messy ...

It comes from the requirement that submessage headers in RTPS message are aligned to a multiple of 4 bytes and that data payloads start at an multiple of 4 bytes as well, which means that data (may) need to be padded out to a multiple of 4 bytes, too.

For better or for worse, I chose to include that padding in the serialized representation in the `serdata` in the C version, from where it metastasised to C++ and Python. The idea was that this allows a DATA submessage to be a pair of blobs in `sendmsg` (one containing just the message header, one the payload), instead of a triplet of blobs (a 3rd one for padding). Of course there are other options: in hindsight, it would've been better to *prepend* 4 bytes to the message header, because that padding at the end of the payload is highly annoying and trips everyone up — that's including me 😳😱🤦

(I think I can actually still make that change without causing any problems for a `serdata` implementation that does pad it out. It would just eliminate the need to do so.)

Anyway, the only place where it matters is in sending the data out on the wire. The "lookup instance" operation by definition doesn't do that, therefore it can't *require* the padding. It does make some sense to just pad everything out in the `pub` side of things, of course.

The keys do need to be the same because those get compared, but the Python binding (like the C code) stores a serialised key separately from the data, and it obtains it by calling `dds_stream_extract_key_from_key` or `dds_stream_extract_key_from_data`.

I think it might make the most sense to leave the padding bit here as it is now, and eliminate this weird padding instead.